### PR TITLE
Fixed precondition script to run on win

### LINKED ISSUE #265: Error msg when running toolkit on Windows
when i ran it on Windows, I got the following error: 

module.service_project.module.project-factory.null_resource.preconditions (local-exec): Executing: ["cmd" "/C" ".terraform/modules/service_project/modules/core_project_factory/scripts/preconditions.sh \\\r\n    --credentials_path '' \\\r\n    --impersonate_service_account '' \\\r\n    --billing_account 'xxxxx' \\\r\n    --org_id 'xxxx' \\\r\n    --folder_id 'folders/xxxx' \\\r\n    --shared_vpc 'xxx'\r\n"]
module.service_project.module.project-factory.null_resource.preconditions (local-exec): '.terraform' is not recognized as an internal or external command,
module.service_project.module.project-factory.null_resource.preconditions (local-exec): operable program or batch file.

Error: Error running command '.terraform/modules/service_project/modules/core_project_factory/scripts/preconditions.sh \
    --credentials_path '' \
    --impersonate_service_account '' \
    --billing_account 'xxxx \
    --org_id 'xxxx' \
    --folder_id 'folders/xxxx' \
    --shared_vpc 'xxxx'
': exit status 1. Output: '.terraform' is not recognized as an internal or external command,
operable program or batch file.

### PR DESCRIPTION
**Pull request by paulpalamarchuk**
* Fix #265

### CODE REVIEW

**Comment by paulpalamarchuk** [comment]
Running on windows:
```
...
module.project-factory.null_resource.shared_vpc_subnet_invalid_name[0]: Creating...
module.project-factory.null_resource.preconditions: Creating...
module.project-factory.random_id.random_project_id_suffix: Creating...
module.project-factory.null_resource.preconditions: Provisioning with 'local-exec'...
module.project-factory.null_resource.shared_vpc_subnet_invalid_name[0]: Creation complete after 0s [id=2878104276274484690]
module.project-factory.null_resource.preconditions (local-exec): Executing: ["pip3" "install" "-r" "C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt"]
module.project-factory.random_id.random_project_id_suffix: Creation complete after 0s [id=6n8]
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: google-api-python-client~=1.7 in c:\python\lib\site-packages (from -r 
C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt (line 1)) (1.7.11)
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: google-auth~=1.6 in c:\python\lib\site-packages (from -r C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt (line 2)) (1.6.3)
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: httplib2<1dev,>=0.9.2 in c:\python\lib\site-packages (from google-api-python-client~=1.7->-r C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt (line 1)) (0.14.0)
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: google-auth-httplib2>=0.0.3 in c:\python\lib\site-packages (from google-api-python-client~=1.7->-r C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt (line 1)) (0.0.3)
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: six<2dev,>=1.6.1 in c:\python\lib\site-packages (from google-api-python-client~=1.7->-r C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt 
(line 1)) (1.12.0)
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: uritemplate<4dev,>=3.0.0 in c:\python\lib\site-packages (from google-api-python-client~=1.7->-r C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt (line 1)) (3.0.0)
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: rsa>=3.1.4 in c:\python\lib\site-packages (from google-auth~=1.6->-r C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt (line 2)) (4.0)
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: pyasn1-modules>=0.2.1 in c:\python\lib\site-packages (from google-auth~=1.6->-r C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt (line 2)) (0.2.7)
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: cachetools>=2.0.0 in c:\python\lib\site-packages (from google-auth~=1.6->-r C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt (line 2)) (3.1.1)
module.project-factory.null_resource.preconditions (local-exec): Requirement already satisfied: pyasn1>=0.1.3 in c:\python\lib\site-packages (from rsa>=3.1.4->google-auth~=1.6->-r C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/requirements.txt (line 2)) (0.4.7)
module.project-factory.null_resource.preconditions (local-exec): WARNING: You are using pip version 19.2.3, however version 19.3.1 is available.
module.project-factory.null_resource.preconditions (local-exec): You should consider upgrading via the 'python -m pip install --upgrade pip' command.
module.project-factory.null_resource.preconditions: Provisioning with 'local-exec'...
module.project-factory.null_resource.preconditions (local-exec): Executing: ["cmd" "/C" "python C:/.../terraform-google-project-factory/modules/core_project_factory/scripts/preconditions/preconditions.py --billing_account 01B29E-828C10-4D001A --org_id 684124036889 --credentials_path C:\\...\\apps\\terraform-google-project-factory\\credentials.json   "]
module.project-factory.null_resource.preconditions: Still creating... [10s elapsed]
module.project-factory.null_resource.preconditions: Creation complete after 14s [id=3913477512579808351]
module.project-factory.google_project.main: Creating...
...
```

File: modules/core_project_factory/main.tf [hunk 1]
```diff
@@ -74,19 +74,14 @@ resource "null_resource" "preconditions" {
...
-    --impersonate_service_account '${var.impersonate_service_account}' \
-    --billing_account '${var.billing_account}' \
-    --org_id '${var.org_id}' \
-    --folder_id '${var.folder_id}' \
-    --shared_vpc '${var.shared_vpc}'
-EOD
-
-
+    command     = local.pip_requirements_absolute_path
+    interpreter = ["pip3", "install", "-r"]
```
[the remark this answers is not in the document]
Comment (paulpalamarchuk) [reply modules/core_project_factory/main.tf]: It depends on installation way. Basically - it does not.

File: modules/core_project_factory/locals.tf [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  root_path                        = abspath(path.root)
+  pip_requirements_absolute_path   = join("/", [local.root_path, path.module, "scripts/preconditions/requirements.txt"])
+  preconditions_py_absolute_path   = join("/", [local.root_path, path.module, "scripts/preconditions/preconditions.py"])
+  billing_account_attr             = var.billing_account != "" ? "--billing_account ${var.billing_account}" : ""
```
[the remark this answers is not in the document]
Comment (paulpalamarchuk) [reply modules/core_project_factory/locals.tf]: reworked

File: modules/core_project_factory/main.tf [hunk 3]
```diff
@@ -74,19 +74,14 @@ resource "null_resource" "preconditions" {
...
+    on_failure  = "continue"
+  }
+  provisioner "local-exec" {
```
Comment (morgante) [inline modules/core_project_factory/main.tf]: You should always have a newline in between resources.

File: modules/core_project_factory/locals.tf [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  preconditions_py_absolute_path = join("/", [local.preconditions_path, "preconditions.py"])
+  attributes = {
+    billing_account_attr             = var.billing_account
+    org_id_attr                      = var.org_id
+    credentials_path_attr            = var.credentials_path
+    impersonate_service_account_attr = var.impersonate_service_account
+    folder_id_attr                   = var.folder_id
+    shared_vpc_attr                  = var.shared_vpc
+  }
+  preconditions_command = join(" ", concat([var.python_interpreter_path, local.preconditions_py_absolute_path], [for key, value in local.attributes : ((value != "") ? join(" ", [join("", ["--", key]), local.attributes[key]]) : "")]))
```
Comment (morgante) [inline modules/core_project_factory/locals.tf]: This is way more complicated than it needs to be. Please separate into different lines and don't use `join` excessively.

File: modules/core_project_factory/main.tf (hunk 3 above)
Comment (paulpalamarchuk) [reply modules/core_project_factory/main.tf]: Fixed

File: modules/core_project_factory/locals.tf (hunk 4 above)
Comment (paulpalamarchuk) [reply modules/core_project_factory/locals.tf]: @morgante I've reworked it. Looks pretty simple in one-line. If It still needs to be separated into different lines, let me know, please.

**Comment by taylorludwig** [comment]
@aaron-lane I came across this PR while working on #336. 

I started adding a different gcloud path to the precondition

But it looks like the precondition hasn't worked after this PR and only passed tests since the provisioner steps is set as `on_failure = "continue"`

Looking at the [last test results](https://console.cloud.google.com/cloud-build/builds/53622202-f499-4970-9fa7-6ed51d45a1d3?project=cloud-foundation-cicd&angularJsUrl=%2Fcloud-build%2Fbuilds%2F53622202-f499-4970-9fa7-6ed51d45a1d3%3Fproject%3D2843445864&authuser=5) you can see in the output 

```sh
Step #2 - "converge": module.project-factory.module.project-factory.null_resource.preconditions (local-exec): Executing: ["/bin/sh" "-c" "python3 /workspace/test/fixtures/minimal/../../../modules/core_project_factory/scripts/preconditions/preconditions.py --billing_account_attr 01D904-DAF6EC-F34EF7 --credentials_path_attr --folder_id_attr folders/435275639291 --impersonate_service_account_attr --org_id_attr 943740911108 --shared_vpc_attr "]
Step #2 - "converge": module.project-factory.module.project-factory.null_resource.preconditions (local-exec): Unable to import Google API dependencies, skipping GCP precondition checks!
Step #2 - "converge": module.project-factory.module.project-factory.null_resource.preconditions: Creation complete after 1s [id=8861976462080598418]```
``` 

A few issues

**1. Dependencies** 

It fails on dependency imports `Unable to import Google API dependencies`
I replicated it locally and tracked it back to a missing `requests` package. 

**2. Attributes**

All the attributes are passed in with a trailing `_attr` as you can see above (eg `billing_account_attr`) but the python script errors out with those `preconditions.py: error: unrecognized arguments: --credentials_path_attr ...`. 

The previous implementation of this passed them in correctly  

**3. Leftover sh script**

The original [sh implementation](https://github.com/terraform-google-modules/terraform-google-project-factory/blob/master/modules/core_project_factory/scripts/preconditions.sh) is no longer used since python check is called directly in terraform but the script was leftover. 

On top of that PR https://github.com/terraform-google-modules/terraform-google-project-factory/pull/331 was just merged in to add a check to the sh script - but is never actually run anywhere.

**Comment by morgante** [comment]
@taylorludwig Could you open an issue to track those problems?

**Comment by taylorludwig** [comment]
@morgante done - #342

**Pull request merged**